### PR TITLE
WS : Handle PATCH requests for attachments

### LINKED ISSUE #28748: [WS] Handle PATCH requests for attachments
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

> Originaly posted by @khouloudbelguith here https://github.com/PrestaShop/autoupgrade/pull/479#pullrequestreview-1004678171

Following the merge of https://github.com/PrestaShop/PrestaShop/pull/27952 it is now possible to use PATCH for API calls.
But when we try to update an attachment (= attached file) with PATCH method, the file is not updated.

PR https://github.com/PrestaShop/PrestaShop/pull/28747 fixes the problem

Attachments are files a user can upload to prestashop.

https://user-images.githubusercontent.com/16067358/173408235-586e989d-3fa5-4ac0-87b4-8456cc030f34.mp4

### Expected behavior

POST and PUT methods work as expected, so when using PATCH http method with the webservice to update an attachment, the file should be correctly updated.

### Steps to reproduce

1. Go to BO > Catalog > Files > Add a new file with type jpeg for example
2. Go to BO > Advanced Parameters > Webservice page & select Patch for the attachment
3. Using postman, I tried to update the attachment with the Patch method
4. See error > the file is not updated

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

7.2

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by nsorosac**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | https://github.com/PrestaShop/autoupgrade/pull/479#pullrequestreview-1004678171
| Type?             | bug fix
| Category?         | WS
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?      | Fixes #28748
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/479
| How to test?      | https://github.com/PrestaShop/autoupgrade/pull/479#pullrequestreview-1004678171

### CODE REVIEW

**Comment by hibatallahAouadni** [comment]
Ping @nsorosac thanks for your PR :rocket: 
Could you please change your base branch from `develop` to `8.0.x` :pray: 

Thanks in advance!

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Could you target the `8.0.x` branch, please ?

**Review by kpodemski — DISMISSED** [review]
Thank you for your PR @nsorosac . Could you rebase it? There are some commits that are not relevant to your change.

File: classes/webservice/WebserviceSpecificManagementAttachments.php [hunk 1]
```diff
@@ -231,6 +232,8 @@ public function manageAttachments()
                     break;
                 case 'PUT':
                     $this->getWsObject()->executeEntityPut();
+                case 'PATCH':
```
Comment (hibatallahAouadni) [inline classes/webservice/WebserviceSpecificManagementAttachments.php:235]: test review

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @nsorosac 

Thanks again for your PR :rocket: 
But, I found a problem, when I try to modify the file of an attachment, it is modified fine :heavy_check_mark: but the file name and extension are saved as a random number, see the screenshot attached:

https://user-images.githubusercontent.com/23308427/178567156-e82ee460-2556-4ea4-b0a5-ccbb309d3fb7.mp4

Please check and feedback.
Thanks!

**Comment by jolelievre** [comment]
Hi @nsorosac 

although it's not exactly the initial scope maybe it's worth fixing this. I guess the web service handles the PATCh like a POST and automatically fills the fields with random values without checking if they are already set. It's fine in a POST context, but for a patch that is supposed to allow partial update only not so much.

**Comment by matks** [comment]
@kpodemski I think we maybe need to continue this PR to deliver a complete PATCH management of webservice request in PS 8.0.0

**Comment by matks** [comment]
I checked the PR of @nsorosac about the latest issue found by QA team.

So what happens is that:
- PATCH HTTP request is made using `binary` input -> the file is simply piped into the http request 😄 
- PrestaShop Webservice handles incoming http request and we enter `executeFileAddAndEdit()` function https://github.com/PrestaShop/PrestaShop/blob/develop/classes/webservice/WebserviceSpecificManagementAttachments.php#L304
- in this function, PrestaShop uses a FileUploader to write the file binary content on the disk
- and the function actually retrieves the filename and filetype of ... the new file, written on disk 😄 
- naturally filename and filetype could not be retrieved from http request as file content was simply piped in as binary

And PrestaShop saves the persisted data, and modify the filename in SQL using the filename and filetype of the file written on disk. Not the one on the client computer.

Only thing I can imagine is transform
```
            $attachment->file = $file['id'];
            $attachment->file_name = $file['file_name'];
            $attachment->mime = $file['mime_type'];
            $attachment->name[Configuration::get('PS_LANG_DEFAULT')] = $_POST['name'] ?? $file['file_name'];
```
into
```
            $attachment->file = $file['id'];
            if ($this->getWsObject()->method !== 'PATCH') {
                $attachment->file_name = $file['file_name'];
                $attachment->mime = $file['mime_type'];
                $attachment->name[Configuration::get('PS_LANG_DEFAULT')] = $_POST['name'] ?? $file['file_name'];
            }
```

Not very pretty 🤷‍♂️

**Comment by matks** [comment]
@nsorosac We want to complete this topic for 8.0.0 that is coming very soon so I fetched your commits and added them on a branch of mine, that I will submit for merge -> https://github.com/PrestaShop/PrestaShop/pull/29638

Your commits are still marked as authored by you and you will be credited ;)

**Comment by matks** [comment]
Hello @nsorosac I close your Pull Request because I completed your work and got it merged in https://github.com/PrestaShop/PrestaShop/pull/29638

but your work and commit was preserved 😉

**Comment by jolelievre** [comment]
Closed in favor of https://github.com/PrestaShop/PrestaShop/pull/29638

**Pull request closed without merging**